### ISSUE TITLE
Feature: Line numbers in tags file (ctags)

### ISSUE THREAD

**Issue by dvquentincarrel** [ibhagwan/fzf-lua#1547]
### Have you RTFM'd?

- [X] I have done proper research

### Feature Request

Hello, I believe the "tags" subcommand can better exploit the content of a tag file to provide meaningful results to the user.  
This is the first issue I open on an open-source project, so please do not hesitate to call out any faux-pas I may have made.

### Context
By default, Universal ctags (and I believe, other ctags variants too) use a regular expression to location a symbol in a code file.   
This can lead to issues if two symbols have the same signature in a single file, where the location of each of the two symbol will be defined by the exact same regular expression, thus preventing from distinguishing them.  
I work in a codebase where this is often the case.  
Ctags allow the usage of a line number instead of a regular expression (universal ctags allows both at once), to indicate the location of a symbol in the file. 

### Issue
I have noticed the handling of line numbers as what is called the `{tagaddress}` in `:h tagsrch.txt` is a bit peculiar in fzf-lua.  
When only the line number are used (`ctags --excmd=number`), fzf-lua always shows the start of the file in the preview, and redirects the user to the first line of the file upon the selection of the tag.  
When both the regex and the line number are included in the tags file (`ctags --excmd=combined`), fzf-lua still seem to ignore the line number.

### Request
I believe fzf-lua would benefit from being able to properly use tags file which use line numbers instead of a regex.  
I have looked and tinkered around, and it seems the implementation of the feature has already been attempted.

For instance, when using both line numbers and regexes in the tags file, replacing the code in the function M.vimcmd_entry_function :
https://github.com/ibhagwan/fzf-lua/blob/d9db3b684384bcfd1393914b76f89f0205f183a3/lua/fzf-lua/actions.lua#L212-L216
by something like:
```lua
elseif entry.ctag then
  if entry.line then
      vim.api.nvim_win_set_cursor(0, { entry.line, 0 })
  else
      vim.api.nvim_win_set_cursor(0, { 1, 0 })
      vim.fn.search(entry.ctag, "W")
  end
```
correctly uses the line number to redirect the user in the file.

However, this does not work if only line numbers are used in the tags file, as the `entry` table would not contain a `ctag` element.  
The pattern extracting the line of a tag (`local line, tag = text:match("(%d-);?(/.*/)")` in make_entry.lua, function `M.tag`) expects a line number to always be followed by a regex. Maybe falling back to attempting to use a line number would be good ?

Regarding the previewer, I managed to have it behave as I would expect by replacing the following code:
https://github.com/ibhagwan/fzf-lua/blob/d9db3b684384bcfd1393914b76f89f0205f183a3/lua/fzf-lua/previewer/builtin.lua#L1127-L1135
with:
```lua
if not entry.line then
  api.nvim_win_set_cursor(0, { 1, 0 })
  fn.clearmatches()
  fn.search(entry.ctag, "W")
  if self.win.hls.search then
    fn.matchadd(self.win.hls.search, entry.ctag)
  end
else
  api.nvim_win_set_cursor(0, { entry.line, 0 })
end
self.orig_pos = api.nvim_win_get_cursor(0)
```
But this feels hackish, and still requires both the line number and regex to be in the tags file.

Do you believe the implementation of this feature would make sense in the project?

**Comment by ibhagwan** [comment]
> I have looked and tinkered around, and it seems the implementation of the feature has already been attempted.

You’re correct, I forgot which command I used to test the ctags with line numbers and I’m not sure if it’s still operational but if a line/col is detected it will be used in the preview.

> But this feels hackish, and still requires both the line number and regex to be in the tags file.

Correct, it is somewhat hacky as I believe there could be different formats with line and it’s not always guaranteed to be in the same spot? I might be wrong it’s been a while.

> Do you believe the implementation of this feature would make sense in the project?

The reason I don’t really like this is due to the nature of tags which are somewhat like symbols and unless you’re constantly regenerating the ctags every added line could mess up the location of your line based tags.

Aside from that there are other alternatives that provide similar functionality that do support lines such as the LSP pickers and also `:FzfLua treesitter` where LSP isn’t supported.

**Comment by dvquentincarrel** [comment]
> You’re correct, I forgot which command I used to test the ctags with line numbers and I’m not sure if it’s still operational but if a line/col is detected it will be used in the preview.

From what I've been able to test, it does not seem to be completely operational, at least anymore. The line appears in the picker in green, but the preview completely ignores it, which is a bummer if two symbols are located by the same regex, but at opposite ends of a file.

> Correct, it is somewhat hacky as I believe there could be different formats with line and it’s not always guaranteed to be in the same spot? I might be wrong it’s been a while.

Based on what I could find in the man pages, only universal-ctags has the option to use both regex and line number at once, all the other can use either one, or the other on any given line.  
I'm not sure what "it" you're referring to when you say "it's not always guaranteed to be in the same spot". If you mean code, indeed not. If you mean line number, inside of the line of the tag file, I *think* it actually is.

> The reason I don’t really like this is due to the nature of tags which are somewhat like symbols and unless you’re constantly regenerating the ctags every added line could mess up the location of your line based tags.

You're right, and I agree more modern alternatives are better *when they're available*.

The reason I turned to tags is because the codebase I work on has some *quirks*. One of which being that the framework we use attempts to mock a prototypal kind of class inheritance in a language that very much does not do prototypal inheritance. Parent classes to inherit from are also passed around as strings and looked up dynamically.  
Stock LSPs are unfortunately unable to handle this kind of *creative thinking*, but tags provide a rather convenient and robust way to quickly search for something defined somewhere else in the codebase.
I personally have a cron triggered every 5 minutes that regenerates tags for the whole project (~2mil LoC, takes barely a couple of seconds). I use the tags for codebase traversal in conjunction with LSP pickers for the current file I'm editing, and it works alright.

I think providing *something* for the case where the tag file does not use regexes at all could prove beneficial since it just does not work right now, even if the behavior of regex+line numbers is kept the same, but I understand your apprehension towards it.  
At the end of the day, the modifications presented above allow it to work with my use case, I just thought It could be nice to allow other people to benefit from it :)

This is off topic, but we also throw away what are technically comments in the tag line, but which is where the metadata is actually stored (the kind of the symbol (class, variable, function, ...), which class does it belong to, etc...).
I assume adding an option to have it in the picker line, to allow more fine-grained selection, is not something you'd consider interesting given how much tags have been supplanted by LSPs and the likes ?

**Comment by ibhagwan** [comment]
> From what I've been able to test, it does not seem to be completely operational, at least anymore. The line appears in the picker in green, but the preview completely ignores it, which is a bummer if two symbols are located by the same regex, but at opposite ends of a file.

Are you using bat previewer or the default?
In any event I would consider this a bug that should be fixed.

> I think providing something for the case where the tag file does not use regexes at all could prove beneficial since it just does not work right now, even if the behavior of regex+line numbers is kept the same, but I understand your apprehension towards it.
At the end of the day, the modifications presented above allow it to work with my use case, I just thought It could be nice to allow other people to benefit from it :)

I’m with you on that, if we can make both cases with and without lines work consistently (as it once was) - why not?

> This is off topic, but we also throw away what are technically comments in the tag line, but which is where the metadata is actually stored (the kind of the symbol (class, variable, function, ...), which class does it belong to, etc...).
I assume adding an option to have it in the picker line, to allow more fine-grained selection, is not something you'd consider interesting given how much tags have been supplanted by LSPs and the likes ?

last time I modified the ctags format was to equalize it with the fzf.vim formatting which is what users coming from vim are used to.

I’m pretty sure we can “unhide” fields using `- -nth` but I’m answering from my phone so I haven’t looked at the code and therefore might not be as coherent :)

In any event, since you have it mostly fixed, would you like to submit a PR for this, otherwise I can fix it.

**Comment by dvquentincarrel** [comment]
> Are you using bat previewer or the default?

I'm using bat.

> I’m pretty sure we can “unhide” fields using --nth but I’m answering from my phone so I haven’t looked at the code and therefore might not be as coherent :)

Unfortunately, I don't believe we can in this case, since those fields seem to be stripped when the entry in the picker is constructed here because they're technically comments:
https://github.com/ibhagwan/fzf-lua/blob/d9db3b684384bcfd1393914b76f89f0205f183a3/lua/fzf-lua/make_entry.lua#L429-L433

I think we would have to add an option to specifically not throw away the comments, and append them at the end of the entry if we were to add support for them.

I can submit a PR for it, should be done in a couple of hours or tomorrow :)

**Comment by ibhagwan** [comment]
Resolved with #1550.
